apiserver/facades/agent/controllercharm/controllercharm.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package controllercharm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+
+	apiservererrors "github.com/juju/juju/apiserver/errors"
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/environs/bootstrap"
+	"github.com/juju/juju/rpc/params"
+	"github.com/juju/juju/state"
+)
+
+const (
+	// jujuMetricsUserPrefix defines the "namespace" in which this facade is
+	// allowed to create/remove users.
+	jujuMetricsUserPrefix = "juju-metrics-"
+
+	// userCreator is the listed "creator" of metrics users in state.
+	userCreator = "admin"
+)
+
+// API provides API methods to the controllercharm worker.
+type API struct {
+	state backend
+}
+
+// backend defines the state methods that API needs.
+type backend interface {
+	AddUser(name string, displayName string, password string, creator string) (*state.User, error)
+	RemoveUser(tag names.UserTag) error
+	Model() (model, error)
+}
+
+// model defines the model methods that API needs.
+type model interface {
+	AddUser(state.UserAccessSpec) (permission.UserAccess, error)
+}
+
+// stateShim allows the real state to implement backend.
+type stateShim struct {
+	*state.State
+}
+
+func (s stateShim) Model() (model, error) {
+	return s.State.Model()
+}
+
+// AddMetricsUser creates a user with the given username and password, and
+// grants the new user permission to read the metrics endpoint.
+func (api *API) AddMetricsUser(args params.AddUsers) (params.AddUserResults, error) {
+	var results params.AddUserResults
+	for _, user := range args.Users {
+		err := api.addMetricsUser(user)
+		results.Results = append(results.Results, params.AddUserResult{
+			Tag:   user.Username,
+			Error: apiservererrors.ServerError(err),
+		})
+	}
+	return results, nil
+}
+
+func (api *API) addMetricsUser(args params.AddUser) error {
+	if !strings.HasPrefix(args.Username, jujuMetricsUserPrefix) {
+		return errors.NotValidf("username %q missing prefix %q", args.Username, jujuMetricsUserPrefix)
+	}
+
+	if args.Password == "" {
+		return errors.NotValidf("empty password for user %q", args.Username)
+	}
+
+	_, err := api.state.AddUser(args.Username, args.DisplayName, args.Password, userCreator)
+	if err != nil {
+		return errors.Annotatef(err, "failed to create user %q", args.Username)
+	}
+
+	model, err := api.state.Model()
+	if err != nil {
+		return errors.Annotatef(err, "retrieving current model")
+	}
+
+	userTag := names.NewUserTag(args.Username)
+	_, err = model.AddUser(state.UserAccessSpec{
+		User:      userTag,
+		CreatedBy: names.NewUserTag(userCreator),
+		Access:    permission.ReadAccess,
```

Comment:
Probably consider adding a new permission for metric endpoint only.

Reply:
We have discussed this, and my current thinking is it's not worth the squeeze. If someone has access to the metrics endpoint, they already have access to all the metrics from the Juju controller and all its models. So they already know **a lot** about the Juju controller, in which case, we might as well grant them `read:controller`. In other words, there's not enough difference between `read-metrics` and `read` to justify adding the new permission.

Interested to get a take from @hpidcock though.

Reply:
I think this would be nice if it was a permission system, but is not, IIANM this is an access level system, i.e. the levels are comparable and include the "permissions" of the previous level. I think read access is probably fine.

Reply:
yeah, from what I've seen, it would be really difficult to change the current permissions model. It would be nice to have more expressive permissions in Juju but until then, we can just stick with read:controller here.

Reply:
Had a discussion today, we're going to leave the permissions as they are for now. We can revisit this later.